app/assets/javascripts/controllers/courses/controller.js

```diff
@@ -9,47 +9,192 @@ Yacs.views.courses = function (target, params) {
   params.show_sections = params.show_periods = true;
 
   /**
-   * When a section is clicked, toggle whether it is selected.
-   * Uses Yacs.user for POT of selections
+   * When a section is clicked, toggle its individual selection status.
    * When a course is clicked, toggle its sections as selected.
-   * Uses CSS for POT of selections
+   * Uses Yacs.user as the point of truth for selections
    */
   var bindListeners = function () {
     Yacs.on('click', target.querySelectorAll('section'), function (section) {
-      Yacs.user.removeSelection(section.dataset.id) ||
-        Yacs.user.addSelection(section.dataset.id);
+      var sid = section.dataset.id;
+      var cid = section.dataset.courseId;
+      if(! Yacs.user.removeSelection(sid, cid)) {
+        // remove failed so add instead
+        Yacs.user.addSelection(sid, cid, true);
+      }
     });
 
     Yacs.on('click', target.querySelectorAll('course-info'), function (courseInfo) {
-      var courseSelected = courseInfo.parentElement.classList.contains('selected');
-      if (courseSelected) {
-        var sections = courseInfo.parentElement.querySelectorAll('section');
-        Yacs.user.removeSelections(map(sections, function (section) {
-          return section.dataset.id;
-        }));
+      var cid = courseInfo.parentElement.dataset.id;
+      if (Yacs.user.courseIsSelected(cid)) {
+        Yacs.user.removeCourse(cid);
       } else {
         var sections = courseInfo.parentElement.querySelectorAll('section:not(.closed)');
-        Yacs.user.addSelections(map(sections, function (section) {
+        // TODO: optimize
+        Yacs.user.addMultipleSelections(map(sections, function (section) {
           return section.dataset.id;
-        }));
+        }), cid);
+      }
+    });
+  };
+
+  /**
+   * Given the data returned from the API, populate the conflicts cache with
+   * the data for conflicts.
+   */
+  var populateConflictsCache = function(data) {
```

Comment:
Another note: the whole frontend cache shouldn't be scrapped. It'll let us do things like easily show the names of conflicting courses (see #164).